docs/howtos/uniffi-object-destruction-on-kotlin.md

```diff
@@ -0,0 +1,37 @@
+
+## You can create objects in a function if you also destroy them there
+
+The simplest way to get destruction right is to create an object and destroy it in the same function.  The [use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/use.html) function makes this really easy:
+
+```
+SomeUniFFIObject()
+  .use { obj ->
+      obj.doSomething()
+      obj.doSomethingElse()
+  }
+```
+
+## You can create and store objects in singletons
+
+If we are okay with UniFFI objects living for the entire application lifetime, then they can be stored in singletons.  This is how we handle our database connections, for example [SyncableLoginsStorage](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt#L98-L114) and [PlacesReaderConnection](https://github.com/mozilla-mobile/firefox-android/blob/9de5ab8cc098674aad5190ce8b00ae6be65e9bd0/android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Connection.kt#L75-L96).
+
+## You can create and store objects in an class, then destroy them in a lifecycle method
+
+UniFFI objects can stored in classes like the Android [Fragment](https://developer.android.com/guide/fragments/lifecycle) class that have a defined lifecycle, with methods called at different stages.  The class just needs to ensure that the UniFFI object is destroyed in one of the lifecycle methods, for example `Fragement.onDestroy()`.
+
+## You can "borrow" objects
+
+Classes can borrow objects, in the Rust sense of the word, by inputting them in their constructor then storing them.  A good example is the [GeckoLoginStorageDelegate](https://github.com/mozilla-mobile/firefox-android/blob/009fb6350072b8b4cf2b87bd1a8f497843f67843/android-components/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt#L46-L49).  These classes are not required to call the destructor, instead the code that constructs them is responsible for ensuring that the borrowing objects don't outlived the owned objects.
+
+Unlike in Rust, there's no borrow checker to ensure that code follows these rules.  If we make a mistake, then the borrowing class may try to use the object after it's already been destroyed.  In this case the generated code will raise an `IllegalStateException`.  This should obviously be avoided, but at least it won't result in memory corruption.
+
+## Avoid classes that take ownership of objects.
+
+The alternative to borrowing is taking ownership of objects: Classes that take ownership of the objects they input.  This is how the UniFFI generated code works, for example UniFFI Records/Enums that contain UniFFI objects define their own destroy() method that calls destroy on the child object.
```

Comment:
I'm not quite sure what this section is trying to say - and isn't it true that taking ownership would actually be fine, so long as you manage that ownership correctly?

IOW, it seems this section and the above "borrow" section are really just saying "you must manage lifetimes correctly. You can "borrow" or you can pass ownership around, but it's on you if you get it wrong - and if you do, you'll get `IllegalStateException` which is never going end well, so don't do that" :)